### ISSUE TITLE
Multi-thread upload for different backends

### ISSUE THREAD

**Issue by NyaMisty** [rclone/rclone#4798]
#### What is your current rclone version (output from `rclone version`)?
rclone v1.53.3 :)

#### What problem are you are trying to solve?

Currently, backend can only upload file in one thread, or do some buffering and implement multi thread in backend's own logic.

If we use only one thread, when faced with super big file (like 700GB~2TB), the uploading would be very unstable and slow due to TCP flow control & various QoS.

But if we implement buffering in backend, that'll mean each backend would need to implement their own multi thread management logic, which I think is redundant and not that elegant.

#### How do you think rclone should be changed to solve that?

For most backend, they split the file into chunks of various size before uploading. So, apparently we should implement multi-thread transfer by directly offering chunks to backend (block-level sequential access), instead of using OpenWriterAt interface (totally random access), which could seldomly be implemented on a remote fs.
So personally I suggest we add a special Put interface to receive a ChunkGetter, which reads chunks from the source file object in a multi-threaded manner, and then reimplement a multi-thread-streams for uploading process.

**Comment by ncw** [comment]
Yes there is a plan for this, mostly for enabling resume of large file transfers but I think it would work for your purpose.

See #87

**Comment by NyaMisty** [comment]
Thanks for you reply! 
It seems that [some member](https://github.com/rclone/rclone/issues/87#issuecomment-717456301) thinks this should be done by chunker / Resumer interface, and some discussion on the forum believe that this should be implemented with a special OpenWriterAt method.
For me, I've already implemented a patch for chunker which parallelized the chunk upload process. But I think these ideas are either not multithread-friendly, or hard to implement :(
If we build a interface for chunking, and add a Put2 method into Fs interface, maybe the resumable upload will be easier to implement (as this will enable putting slices of file into remote)

And the last one, should I continue the discussion here or re-post my idea to #87 ?

**Comment by ncw** [comment]
I was thinking about a general purpose "multipart streaming upload" function for rclone recently which is very similar to what you are talking about it now occurs to me.

Rclone has the capability of doing random reads from (nearly all) backends.

If a backend has the capability of uploading in multiple streams then the most efficient way to upload would be to read multiple streams at once and upload these multiple streams.

So say we have a 1TB file and are uploading using 4 streams.

We set one thread to download 0-250G and upload that to a chunk on the backend, another thread to download 250-500G and upload that etc.

For somethng like S3 multipart upload, the chunks can be as large as you like and uploaded in parallel.

For other backends the chunks have a maximum size.

So hand-waving a few details this would enable you to stream multiple reads with multiple writes increasing the bandwidth greatly.

I think this would work better than the current chunking system, though they are actually equivalent provided we don't buffer the chunks in memory.

> And the last one, should I continue the discussion here or re-post my idea to #87 ?

Let's continue here. The resume interface is currently too specialized for doing multi-thread uploads (though I think a previous version wasn't

```
Optional interface for backends

type Resumer interface {
    // Resume checks whether the (remote, ID) pair is valid and returns
    // the point the file should be resumed from or an error.
    Resume(ctx context.Context, remote string, ID string) (Pos int64, err error)
}

A new fs.Option for Put/Upload for doing the resuming - OptionResume
should not be passed in unless the backend supports the optional
Resume method.

struct OptionResume {
    ID string             // resume this ID if set
    Pos int64             // and resume from this position
    SetID func(ID string) // called when and if the upload knows an ID for resumable upload
}

operations.Copy would initially set just the SetID function pointer -
it would then call Put/Update and it would get called back with an ID
if a multipart upload was in use.

On the callback it would persist the ID along with the Fs, path and
the Fingerprint of the source object.

If OptionResume is passed in then the backend should not cancel
multipart uploads on failure.

If Put/Upload returned OK then the upload was successful so the on
disk metadata can be removed.

So when being re-run operations.Copy looks on disk for uncompleted
uploads for this (Fs, remote, ID) and checks that the fingerprint of
the source object hasn't changed.

If the fingerprint is unchanged then it calls the Resume method
If not then it calls this optional
method

If this returns an error then Copy carries on as usual.
```

For multi thread uploads we'd want another call to set up the upload. This would take parameters

in
- size of file (if known) - not all backends can upload a file whose size isn't known
- desired number of upload threads

out
- ID (maybe for use with ResumeOption)
- Chunk size to use - 0 for unlimited?
- Total number of chunks - ideally would be same as desired number of upload threads, but could be more
- Boolean as to whether chunks can be uploaded in parallel or not - not all backends can upload in parallel

If done right then this might be able to get rid of the multipart upload code in all the backends (a lot of which is very similar).

I think that needs a bit more thought, but there are some ideas!

> For most backend, they split the file into chunks of various size before uploading. So, apparently we should implement multi-thread transfer by directly offering chunks to backend (block-level sequential access), instead of using OpenWriterAt interface (totally random access), which could seldomly be implemented on a remote fs.
> So personally I suggest we add a special Put interface to receive a ChunkGetter, which reads chunks from the source file object in a multi-threaded manner, and then reimplement a multi-thread-streams for uploading process

I think we are thinking on the same lines. What would your suggested interface be?

**Comment by ivandeex** [comment]
Links to note.
While this issue seeks to find a _common_ base for threaded operations (thus marked with _thinking_), https://github.com/rclone/rclone/pull/4821 tries to increase concurrency in one place and https://github.com/rclone/rclone/pull/4172 hurdles it in the other...

**Comment by ivandeex** [comment]
Marking concurrency related stuff by a label for better visibility.

**Comment by ivandeex** [comment]
@ncw @NyaMisty 
Some cloud servers in fact support partial file write.
For example, see SSHFS code: https://github.com/libfuse/sshfs/blob/master/sshfs.c#L3231
Having partial writes will let me use rclone FUSE/VFS to `mmap` a large file and update it in real time.
My use-case is a large `bolt` database (> 20G), bolt internally does mmap and performs pretty small scattered read/write accesses.

Translation of `mmap`-ed page accesses to ranged read/writes is performed for us by the kernel fuse driver.
I just want that rclone translate it into most effective transfers supported by backend.
SSHFS can do that, but rclone currently can upload only full files.

We need a new internal interface to have this feature.
@ncw @NyaMisty Can the API discussed here be used for that?
Thanks

**Comment by ncw** [comment]
@ivandeex I think you want a different interface here...

This is much more like the interface you want (only the local backend implements this at the moment).

```
        // OpenWriterAt opens with a handle for random access writes
        //
        // Pass in the remote desired and the size if known.
        //
        // It truncates any existing object
        OpenWriterAt func(ctx context.Context, remote string, size int64) (WriterAtCloser, error)
```

Note that quite a few cloud storage backends can write to the middle of files in a round about way. For example with s3 (and b2) you can copy segments out of existing files into a new one. So if you wanted to upload a new chunk in the middle of an existing object, you'd upload the chunk then copy the start of the old object, the new chunk, then the end of the old object over the existing object. These operations are rather "expensive" in terms of transactions and time though. I haven't figured out a sensible way for rclone to use these though.

**Comment by ivandeex** [comment]
OK, I will try to implement OpenWriterAt for the SFTP backend
and make VFS use it with `vfs-cache-mode=writes` or `vfs-cache-mode=none`.
This should let user (me) mmap large files mounted by rclone and perform small changes without need to reserve large space on disk for cache.
References:
- https://github.com/rclone/rclone/blob/v1.56.0/backend/local/local.go#L1228
- https://github.com/rclone/rclone/blob/v1.56.0/fs/operations/multithread.go#L142
- https://github.com/rclone/rclone/blob/v1.56.0/fstest/fstests/fstests.go#L738

**Comment by ivandeex** [comment]
@ncw
OK, I will try to implement OpenWriterAt for the SFTP backend
and make VFS use it with `vfs-cache-mode=writes` or `vfs-cache-mode=none`.
This should let user (me) mmap large files mounted by rclone and perform small changes without need to reserve large space on disk for cache.
References:
- https://github.com/rclone/rclone/blob/v1.56.0/backend/local/local.go#L1228
- https://github.com/rclone/rclone/blob/v1.56.0/fs/operations/multithread.go#L142
- https://github.com/rclone/rclone/blob/v1.56.0/fstest/fstests/fstests.go#L738

**Comment by ncw** [comment]
This has been implemented in #7056 :-)

I'll close this now.
